nav2_safety_nodes/src/nav2_safety_node.cpp

```diff
@@ -0,0 +1,232 @@
+    speed_limit_topic_, rclcpp::QoS(10));
+  // Multiple Laserscan subscribers
+  scan_subscribers_.reserve(scan_topics_.size());
+  RCLCPP_INFO(logger_, "Subscribing to scan topics");
+  for (unsigned int i = 0; i < scan_topics_.size(); i++) {
+    scan_subscribers_.push_back(
+      create_subscription<sensor_msgs::msg::LaserScan>(
+        scan_topics_[i].c_str(), rclcpp::SystemDefaultsQoS(),
+        std::bind(&SafetyZone::laserCallback, this, std::placeholders::_1)));
+  }
+  // Timer -> 10hzs
+  auto update_frequency_s = std::chrono::duration<double>(update_frequency_);
+  auto update_frequency_ms = std::chrono::duration_cast<std::chrono::milliseconds>(update_frequency_s);
+  timer_ = create_wall_timer(
+  update_frequency_ms, std::bind(&SafetyZone::timerCallback, this));
+  RCLCPP_INFO(logger_, "Subscribed to scan topics");
+  safety_polygon_pub_->on_activate();
+  speed_limit_pub_->on_activate();
+
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+nav2_util::CallbackReturn
+SafetyZone::on_deactivate(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(logger_, "Deactivating");
+  safety_polygon_pub_->on_deactivate();
+  speed_limit_pub_->on_deactivate();
+  timer_->cancel();
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+nav2_util::CallbackReturn
+SafetyZone::on_cleanup(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(logger_, "Cleaning up");
+  safety_polygon_pub_.reset();
+  speed_limit_pub_.reset();
+  timer_.reset();
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+nav2_util::CallbackReturn
+SafetyZone::on_shutdown(const rclcpp_lifecycle::State &)
+{
+  RCLCPP_INFO(logger_, "Shutting down");
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+void
+SafetyZone::laserCallback(
+  const sensor_msgs::msg::LaserScan::SharedPtr message)
+{
+  // project the laser into a point cloud
+  sensor_msgs::msg::PointCloud2 cloud;
+
+  try {
+    projector_.projectLaser(*message, cloud);
+  } catch (...) {
+    RCLCPP_WARN(logger_, "Unable to project laser into cloud");
+  }
+
+  // Transform the cloud if necessary
+  if (!base_frame_.empty() && cloud.header.frame_id != base_frame_) {
+    try {
+      cloud = tf2_->transform(cloud, base_frame_, tf2::durationFromSec(tf_tolerance_));
+      sensor_data_.push(cloud);
+    } catch (tf2::TransformException & ex) {
+      RCLCPP_WARN(logger_, "Transform failure: %s", ex.what());
+      return;
+    }
+  } else if (cloud.header.frame_id == base_frame_) {
+    sensor_data_.push(cloud);
+  }
+}
+
+double
+SafetyZone::dotProduct(
+  const Eigen::Vector3d & pt1,
+  const Eigen::Vector3d & pt2)
+{
+  return pt1[0] * pt2[1] - pt1[1] * pt2[0];
+}
+
+// Function for detecting if cloud points are inside safety polygon or not
+int
+SafetyZone::detectPoints(
+  const sensor_msgs::msg::PointCloud2 & cloud,
+  const std::vector<geometry_msgs::msg::Point> & safety_zone)
+{
+  int points_inside = 0;
+  int count_same_side_results = 0, on_edge = 0;
```

Comment:
You don't reset count_same_side_results or on_edge between point evaluations, this would only work for a single first point!

Reply:
We need to calculate same_side results for every point in pcl so we need to reset those variables to zero for each point, I think I should move these two lines inside first for loop, right?

Reply:
yes

Reply:
I have done that